distsqlrun,kv,client: don't create flows in aborted txns

andreimatei opened the pull request:
Before this patch, we had a race between a heartbeat find out (async)
that a txn has been aborted and a client using the txn to create DistSQL
flows. We could end up creating flows in aborted txns, which are going
to have a bad time: at the moment, leaf TxnCoordSenders don't react well
to what appears to be an async abort: tcs.maybeRejectClientLocked()
returns a HandledRetryableError, but the DistSQL infrastructure only
expects raw retriable errors - so the HandledRetryableError was
ironically losing its "retryable" character and was making it to a
sql client with the wrong pgwire code.
This patch resolves the situation by atomically checking that the txn is
still good (i.e. status==PENDING) before extracting its metadata that is
sent to leaves.

Fixes #28898
Fixes #29271

Release note: Fix an issue where, under severe load, clients were
sometimes receiving retryable errors with a non-retryable error code (a
client would get an error with the message "HandledRetryableError: ..."
but an internal error code instead of the expected retryable error
code).

andreimatei committed:
distsqlrun: plumb a context

... to a standard callback. Will be used marginally in the next commit.

Release note: None

andreimatei committed:
distsqlrun,kv,client: don't create flows in aborted txns

Before this patch, we had a race between a heartbeat find out (async)
that a txn has been aborted and a client using the txn to create DistSQL
flows. We could end up creating flows in aborted txns, which are going
to have a bad time: at the moment, leaf TxnCoordSenders don't react well
to what appears to be an async abort: tcs.maybeRejectClientLocked()
returns a HandledRetryableError, but the DistSQL infrastructure only
expects raw retriable errors - so the HandledRetryableError was
ironically losing its "retryable" character and was making it to a
sql client with the wrong pgwire code.
This patch resolves the situation by atomically checking that the txn is
still good (i.e. status==PENDING) before extracting its metadata that is
sent to leaves.

Fixes #28898
Fixes #29271

Release note (bug fix): Fix an issue where, under severe load, clients
were sometimes receiving retryable errors with a non-retryable error
code (a client would get an error with the message
"HandledRetryableError: ..." but an internal error code instead of the
expected retryable error code).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/29455)
<!-- Reviewable:end -->

andreimatei commented:
cc @jordanlewis 
Still have to write a test, and I'll add some more protection at different levels for trapping leaves created with aborted txns, but something like this.

nvb reviewed (changes requested):
> Reviewed 14 of 14 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/internal/client/sender.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLWlQb36b6hdtqYl6PJ:-LLWlQb36b6hdtqYl6PK:b3131wl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe617849c0c89591a72bb1dfb2ea4200537ea6f1/pkg/internal/client/sender.go#L109)):*
> ```Go
> 	// AugmentMeta().
> 	//
> 	// If AnyTxnStatus is passed, than this function never returns errors.
> ```

`s/than/then/`
___
*[pkg/sql/distsql_running.go, line 138 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLWlvNP2h8GC-E_CgtD:-LLWlvNP2h8GC-E_CgtE:bc5ar9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe617849c0c89591a72bb1dfb2ea4200537ea6f1/pkg/sql/distsql_running.go#L138)):*
> ```Go
> 		// If the plan is not local, we will have to set up leaf txns using the
> 		// txnCoordMeta.
> 		meta, err := txn.GetStrippedTxnCoordMeta(ctx, client.OnlyPending)
> ```

Mind defending why this approach is better than just checking `meta.Status` here and not complicating every other caller of `GetStrippedTxnCoordMeta`?


<!-- Sent from Reviewable.io -->

andreimatei committed:
client/roachpb: replace txn.GetStrippedTxnCoordMeta

... with a pair of methods on the TxnCoordMeta proto.
This is in preparation of txn.GetTxnCoordMeta() being split into two
flavors in the next commit.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/internal/client/sender.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLWlQb36b6hdtqYl6PJ:-LLaI9yX5C4UGrovHoTx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe617849c0c89591a72bb1dfb2ea4200537ea6f1/pkg/internal/client/sender.go#L109)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/than/then/`
</blockquote></details>

Done.
___
*[pkg/sql/distsql_running.go, line 138 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLWlvNP2h8GC-E_CgtD:-LLaID8o-3jJ-wcH3TeO:bbkj4ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe617849c0c89591a72bb1dfb2ea4200537ea6f1/pkg/sql/distsql_running.go#L138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Mind defending why this approach is better than just checking `meta.Status` here and not complicating every other caller of `GetStrippedTxnCoordMeta`?
</blockquote></details>

As we were discussing, it's cause we want the TCS to be in charge of producing the right error.
Check out the API now - I've moved the "stripped" flavor elsewhere, left the txn with two flavors - one produces errors, one doesn't - and now the `TxnStatusOpt` is just used by the `TCS` API, not by the `Txn` (which is more "client" facing).


<!-- Sent from Reviewable.io -->

andreimatei committed:
kv: move test to more standard tracing infra

Release note: None

andreimatei reviewed (commented):
wrote an integration test. PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-:-LLev5u18nYvkC3EEvum:bb3akc2)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei committed:
distsqlrun: refuse to run flows in non-PENDING txns

... now that the gateway is not trying to create them any more.

Release note: None

andreimatei committed:
client: plumb ctx to txn ctors and boil the ocean

Release note: None

nvb reviewed (changes requested):
> Reviewed 11 of 11 files at r3, 6 of 6 files at r4, 17 of 17 files at r5, 1 of 1 files at r6, 2 of 2 files at r7, 15 of 35 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/internal/client/txn.go, line 138 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWrpW5at6o9CwEMip:-LLfWrpW5at6o9CwEMiq:b1eyd7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/92b6676b92541999f0e7ece40270932cb586e006/pkg/internal/client/txn.go#L138)):*
> ```Go
> 		log.Fatalf(context.TODO(), "attempting to create txn with nil db for Transaction: %s", meta.Txn)
> 	}
> 	if meta.Txn.Status != roachpb.PENDING {
> ```

Careful. What about mixed version clusters?
___
*[pkg/sql/integration_test.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfVZ6k68a468j-VpcW:-LLfVZ6k68a468j-VpcX:b-n1cvvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/sql/integration_test.go#L1)):*
> ```Go
> // Copyright 2018 The Cockroach Authors.
> ```

`sql/integration_test.go` is a useless file name. If this doesn't belong in any other test file then name this something descriptive.
___
*[pkg/sql/integration_test.go, line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWEHaE3PWlEDBu5Aw:-LLfWEHaE3PWlEDBu5Ax:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/sql/integration_test.go#L49)):*
> ```Go
> // in it. We're careful to not use the transaction for anything but running the
> // plan; planning will be performed outside of the transaction.
> func TestDistSQLRunningInAbortedTxn(t *testing.T) {
> ```

Nice test!
___
*[pkg/util/tracing/test_utils.go, line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWHqF5JRtE9SdaB_A:-LLfWHqF5JRtE9SdaB_B:b-hvgqy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/util/tracing/test_utils.go#L28)):*
> ```Go
> 		spMsg := ""
> 		for _, l := range recSp.Logs {
> 			for _, f := range l.Fields {
> ```

Why is this needed? Do we expect callers to match against these fields? If so, should we explain this format in the function comment?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/internal/client/txn.go, line 138 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWrpW5at6o9CwEMip:-LLfZHJr3fRLh6PQDWYO:b29eleq) ([raw file](https://github.com/cockroachdb/cockroach/blob/92b6676b92541999f0e7ece40270932cb586e006/pkg/internal/client/txn.go#L138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Careful. What about mixed version clusters?
</blockquote></details>

but I've also added protection in DistSQLServer, a layer above. So DistSQL can't get here with an aborted txn, and that should be the only case where this happened.
___
*[pkg/sql/integration_test.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfVZ6k68a468j-VpcW:-LLfZZuoEpoWuJiBEFeb:bcgtadz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/sql/integration_test.go#L1)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`sql/integration_test.go` is a useless file name. If this doesn't belong in any other test file then name this something descriptive.
</blockquote></details>

It's not useless. It's the most descriptive I can come up with. I'm not gonna name a file just for one test. You can imagine other "integration tests" coming here. There's also precedent in `kv/integration_test.go` :)
___
*[pkg/sql/integration_test.go, line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWEHaE3PWlEDBu5Aw:-LLfZYxiDwd8bOXVGm1F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/sql/integration_test.go#L49)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Nice test!
</blockquote></details>

Done.
___
*[pkg/util/tracing/test_utils.go, line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWHqF5JRtE9SdaB_A:-LLf_-FV5m2iT03_Rhhn:b-49wfsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/util/tracing/test_utils.go#L28)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why is this needed? Do we expect callers to match against these fields? If so, should we explain this format in the function comment?
</blockquote></details>

well the log message itself is one of these fields


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/integration_test.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfVZ6k68a468j-VpcW:-LLf_PIM83qWvazcWL97:bz0rvwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/sql/integration_test.go#L1)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It's not useless. It's the most descriptive I can come up with. I'm not gonna name a file just for one test. You can imagine other "integration tests" coming here. There's also precedent in `kv/integration_test.go` :)
</blockquote></details>

well, I guess this didn't technically end up being an "integration test" after all, it's mostly a test for the `DistSQLPlanner.Run()` method. Let me see about finding another home.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/integration_test.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfVZ6k68a468j-VpcW:-LLfarfk7BKsaJzwOSwA:b-7mmp82) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/sql/integration_test.go#L1)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well, I guess this didn't technically end up being an "integration test" after all, it's mostly a test for the `DistSQLPlanner.Run()` method. Let me see about finding another home.
</blockquote></details>

moved


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-:-LLfcKyPEI7oNET388Z8:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/internal/client/txn.go, line 138 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWrpW5at6o9CwEMip:-LLfcAlHAGDB7zJSPSQK:b-pzfleo) ([raw file](https://github.com/cockroachdb/cockroach/blob/92b6676b92541999f0e7ece40270932cb586e006/pkg/internal/client/txn.go#L138)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

but I've also added protection in DistSQLServer, a layer above. So DistSQL can't get here with an aborted txn, and that should be the only case where this happened.
</blockquote></details>

:+1: just wanted to make sure you had thought about this.
___
*[pkg/sql/distsqlrun/server.go, line 330 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfc_mVA3lohC7HMB5R:-LLfc_mVA3lohC7HMB5S:bfac2mx) ([raw file](https://github.com/cockroachdb/cockroach/blob/adb01ee8b6b7eb22298895bef42d72227a875b46/pkg/sql/distsqlrun/server.go#L330)):*
> ```Go
> 	if meta := req.TxnCoordMeta; meta != nil {
> 		if !localState.IsLocal {
> 			if meta.Txn.Status != roachpb.PENDING {
> ```

nit: Maybe leave a comment that this can be removed in 2.2, unless you think we should always keep it.
___
*[pkg/sql/integration_test.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfVZ6k68a468j-VpcW:-LLfcGS3EpFUJL_B1t-T:bszm2fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2560e6d8b9b566d6a9e6eef48d1fe21836891012/pkg/sql/integration_test.go#L1)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

moved
</blockquote></details>

Much better.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/distsqlrun/server.go, line 330 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfc_mVA3lohC7HMB5R:-LLfdfWXBXZqcYvnGwE-:b-3av0id) ([raw file](https://github.com/cockroachdb/cockroach/blob/adb01ee8b6b7eb22298895bef42d72227a875b46/pkg/sql/distsqlrun/server.go#L330)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Maybe leave a comment that this can be removed in 2.2, unless you think we should always keep it.
</blockquote></details>

I think we should always keep it. It's pretty basic sanity and acts as documentation too.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/internal/client/txn.go, line 138 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-LLfWrpW5at6o9CwEMip:-LLfe5FBCJDWUihAYVbd:b-y8ar0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/92b6676b92541999f0e7ece40270932cb586e006/pkg/internal/client/txn.go#L138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:+1: just wanted to make sure you had thought about this.
</blockquote></details>

btw, I considered having this return an error instead of crash, but didn't do it as the function is too commonly called and it's not worth introducing the error handling I thought


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29455#-:-LLfkH8X4q5xhrNCHuvJ:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
